src/image/VisitInfo.cc

```diff
@@ -0,0 +1,440 @@
+ *
+ * @param[in] metadata  metadata to get
+ * @param[in] key  key name; the associated value is treated as an angle in degrees, if it exists
+ * @return value of metadata for the specified key, as an lsst::afw::geom::Angle,
+ *   with a value of nan if the key is not present
+ */
+geom::Angle getAngle(daf::base::PropertySet const & metadata, std::string const & key) {
+    return getDouble(metadata, key) * geom::degrees;
+}
+
+/**
+ * Set a specified double in a PropertySet, if value is finite
+ *
+ * @param[in,out] metadata  metadata to set
+ * @param[in] key  name of key to set
+ * @param[in] value  value of key
+ * @return true if item set, false otherwise
+ */
+bool setDouble(daf::base::PropertySet & metadata, std::string const & key, double value,
+        std::string const & comment) {
+    if (std::isfinite(value)) {
+        metadata.set(key, value);
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Set a specified angle in a PropertySet, in degrees, if angle is finite
+ *
+ * @param[in,out] metadata  metadata to set
+ * @param[in] key  name of key to set
+ * @param[in] angle  value of key
+ * @return true if item set, false otherwise
+ */
+bool setAngle(daf::base::PropertySet & metadata, std::string const & key, geom::Angle const & angle,
+        std::string const & comment) {
+    return setDouble(metadata, key, angle.asDegrees(), comment);
+}
+
+/**
+ * Get rotation type as a string to use for a for FITS keyword value, given an enum
+ *
+ * @throw lsst::pex::exceptions::RuntimeError if the enum is not recognized
+ */
+std::string rotTypeStrFromEnum(RotType rotType) {
+    switch(rotType) {
+        case RotType::UNKNOWN:
+            return "UNKNOWN";
+        case RotType::SKY:
+            return "SKY";
+        case RotType::HORIZON:
+            return "HORIZON";
+        case RotType::MOUNT:
+            return "MOUNT";
+    }
+    std::ostringstream os;
+    os << "Unknown RotType enum: " << static_cast<int>(rotType);
+    throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, os.str());
+}
+
+/**
+ * Get rotation type as an enum, given the FITS keyword value string equivalent
+ *
+ * @throw lsst::pex::exceptions::RuntimeError if the string is not recognized
+ */
+RotType rotTypeEnumFromStr(std::string rotTypeName) {
+    if (rotTypeName == "UNKNOWN") {
+        return RotType::UNKNOWN;
+    } else if (rotTypeName == "SKY") {
+        return RotType::SKY;
+    } else if (rotTypeName == "HORIZON") {
+        return RotType::HORIZON;
+    } else if (rotTypeName == "MOUNT") {
+        return RotType::MOUNT;
+    }
+    std::ostringstream os;
+    os << "Unknown RotType name: \"" << rotTypeName << "\"";
+    throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, os.str());
+}
+
+class VisitInfoSchema {
+public:
+    table::Schema schema;
+    table::Key<table::RecordId> exposureId;
+    table::Key<double> exposureTime;
+    table::Key<std::int64_t> tai;
```

Comment:
I wish. Unfortunately double precision isn't accurate enough to maintain full precision. That is why DateTime uses nsecs from the unix epoch internally (and one reason it has a rather narrow valid range). Furthermore, the handling of invalid DateTime works using a special value of nsecs and was designed with afw table persistence in mind.